### DISCUSSION TITLE
Direct Asset URL to Storage (S3, Azure, Cloudinary, etc)

### DISCUSSION THREAD

**Discussion by timstapl** [directus/directus#7612] [Feature Requests]
It would be nice to have a public url setting for images and other assets.

This would be most useful when wanting to access files directly on storage (eg directly from S3, Azure, etc.) Or when using something like Cloudinary.

eg, say you are using an azure blob storage account, when you hit the api you can get the filename_disk for an image field, check that 'storage' is azure (or whatever you named it) and then your consumer can build out an image url that points directly to azure if that's what you desire.  (Nuxt in my case).

Inside of rich text fields you don't really have the same option, you just automatically get a `/assets` url for the image (optionally with a token added.) If your Directus instance is not publicly accessible this can be problematic.

Currently we are either having to programmatically update rich text fields as we pull them, or use the url embed method.

I'd like to see an option to prefix with a public url/host, most likely in the interface options:  

![image](https://user-images.githubusercontent.com/11545149/130894435-bb4ecebb-714f-473b-b93c-39caf81a957a.png)

I'm happy to do the underlying work to implement something like this, but I figured it should be an item of discussion first, especially given this message in the docs:  https://github.com/directus/directus/blob/main/docs/reference/api/assets.md#:~:text=%3A%3A%3A%20warning%20Direct%20File,%3A%3A%3A

The primary goal here for my team is to expose images via a CDN url (cloudinary or similar) without having to have code that manually parses the raw wysiwyg content.

**Comment by benhaynes** [comment]
I'll defer to @rijkvanzanten on this one... but it seems like a good option.

**Comment by rijkvanzanten** [comment]
I'm totally down to add a "public_url" env var per storage adapter, that allows you to enter a template string for how to fetch the original file (something like `STORAGE_SPACES_PUBLIC_URL="https://my-bucket.nyc3.digitaloceanspaces.com/{{ filename_disk }}"`) which can then be returned in the `/files` endpoint for use elsewhere. 

Once that exists, we can add an option to the wysiwyg/markdown interfaces for "Use public URL instead of Directus Assets URL".

> but I figured it should be an item of discussion first, especially given this message in the docs:

Very good point. The main issue with relying on third party links is that authentication and thumbnail generation is no longer possible. We'll have to make sure we communicate that clearly to the user, before they expose their whole file storage publicly without intending to do so. 

Another point we'll probably have to solve for sooner or later is how to support private links to those third party services. IIRC S3 et al have ways to have private files with specific temporary access token embedded in the links. Those might be difficult/impossible to generate properly in this returned url. (though that being said, by calling it _PUBLIC\_URL_, we already explicitly consider the URL to be publicly accessible)

**Reply by timstapl** [reply]
Sounds good, what are the next steps? 

I have a fork and can start work on it, or I can stay out of the way if this is something you would rather implement.

Do we need to create an issue from the discussion?

**Reply by rijkvanzanten** [reply]
> I have a fork and can start work on it, or I can stay out of the way if this is something you would rather implement.

It's all yours!

I think the rough steps are:

* Add a new "STORAGE_<LOCATION>_PUBLIC_URL" environment variable that accepts a template string (eg `https://my-s3-cdn.example.com/{{ filename_disk }}`)
* Add a new transformer in here:
https://github.com/directus/directus/blob/36f796a5f2455f44258457d613aad2416a4efd98/api/src/services/payload.ts#L48
that renders the template URL using the file data, and then returns the full public URL (only on read)
* Add a field in here https://github.com/directus/directus/blob/main/api/src/database/system-data/fields/files.yaml that uses that special transformer to make the API return the actual field

**Reply by timstapl** [reply]
I've made some progress on this, but had a question.

If I add a field in `api/src/database/system-data/fields/files.yaml` then I seem to get errors, even with readonly set to true, unless I also add a migration that creates the column. 

Is there a way to create a virtual field? I've looked around a bit but have not found an example of one yet.

**Reply by timstapl** [reply]
To follow up on this a bit, I tried adding the `no-data` and `alias` options to `special` for this field. That does result in no need for a column in the database, however if there is no public_url column then the transformer I setup in the payload doesn't get invoked (there being no value to invoke it on?)

I'll think about this some more, but any pointers would be appreciated :)

**Reply by rijkvanzanten** [reply]
> however if there is no public_url column then the transformer I setup in the payload doesn't get invoked (there being no value to invoke it on?)

https://github.com/directus/directus/blob/eaa97a9272071ed63e070e218eea2472042ebd3e/api/src/services/payload.ts#L181-L198

Looks like those transformers should run regardless of whether the field has data or not 🤔 What are you calling your transformer? Did you add the name of your transformer to the "special" list of the virtual field?

**Reply by timstapl** [reply]
Here is my new field: 

https://github.com/retrocede/directus/blob/e17fb399a140a971ee8e3d67a0e31a5bc99ffef2/api/src/database/system-data/fields/files.yaml#L140-L146

and here is the transformer:

https://github.com/retrocede/directus/blob/e17fb399a140a971ee8e3d67a0e31a5bc99ffef2/api/src/services/payload.ts#L119-L134

**Reply by rijkvanzanten** [reply]
Shit missed this earlier, sorry. Glad to see you got the PR up though 🙂

**Comment by timstapl** [comment]
Pull Request created: https://github.com/directus/directus/pull/8065

**Reply by rijkvanzanten** [reply]
Legend 🙌🏻

**Reply by SalahAdDin** [reply]
And waiting for this, **Cloudinary** became my main CDN since the first time I worked with a Headless-CMS (**Strapi**), being cheaper and more accessible than Amazon S3.

It makes me far from **Directus**, but all its features are tempting me to use it.

**Reply by sgreiner** [reply]
Would love to see that PR merged, currently my devs need to check the storage: 'azure' field and prepend the CDN domain to the filename_disk.

**Reply by lpigna** [reply]
Hoping this feature is ready soon, we want to implement Directus as the main CMS in our project but it is a requirement from my leaders to expose the storage image url of our S3 bucket🤞🏻.

**Reply by paescuj** [reply]
Hey @lpigna, I'm not an official member but I'd still like to mention that there's the possibility to sponsor a specific feature to increase its priority. See https://github.com/directus/directus/pull/8065#issuecomment-1090416847 (hidden comment) and https://docs.directus.io/getting-started/support.html#sponsored-work.

**Reply by ardenschager** [reply]
I'm also hoping this can become a feature. It would be nice to take advantage of the CDN my storage provider has :) it seems like the URL is stored internally from looking at the source, so it should not be too hard to expose it toolside. As well, it would be nice to get the URLs of the other images Directus likes to generate (thumbnail, auto-resized images, etc.)

**Reply by rijkvanzanten** [reply]
@natschager Question! I just posted this on https://github.com/directus/directus/discussions/17359#discussioncomment-5743028 as well, but seeing this feature request is more or less the same, I'd love to hear your thoughts as well:

What would be your "ideal" workflow for dealing with thumbnails? The /assets endpoint is currently handling permissions checks and the generation of the thumbnail when it doesn't exist. The way I use it in production is by having a CloudFront distribution in front of Directus itself, so you still get the asset performance you're looking for, while keeping the ability to have private files and auto-generated thumbnail sizes. In the OP, it kinda sounds like the idea was to have the assets endpoint auto-redirect to a public URL if the file exists? Just trying to wrap my head around the ideal workflow here :)

---

> Hey @lpigna, I'm not an official member but I'd still like to mention that there's the possibility to sponsor a specific feature to increase its priority. See [#8065 (comment)](https://github.com/directus/directus/pull/8065#issuecomment-1090416847) (hidden comment) and https://docs.directus.io/getting-started/support.html#sponsored-work.

Ha, you are now @paescuj! 😄

**Reply by ardenschager** [reply]
> @natschager Question! I just posted this on [#17359 (comment)](https://github.com/directus/directus/discussions/17359#discussioncomment-5743028) as well, but seeing this feature request is more or less the same, I'd love to hear your thoughts as well:
> 
> What would be your "ideal" workflow for dealing with thumbnails? The /assets endpoint is currently handling permissions checks and the generation of the thumbnail when it doesn't exist. The way I use it in production is by having a CloudFront distribution in front of Directus itself, so you still get the asset performance you're looking for, while keeping the ability to have private files and auto-generated thumbnail sizes. In the OP, it kinda sounds like the idea was to have the assets endpoint auto-redirect to a public URL if the file exists? Just trying to wrap my head around the ideal workflow here :)
> 
> > Hey @lpigna, I'm not an official member but I'd still like to mention that there's the possibility to sponsor a specific feature to increase its priority. See [#8065 (comment)](https://github.com/directus/directus/pull/8065#issuecomment-1090416847) (hidden comment) and https://docs.directus.io/getting-started/support.html#sponsored-work.
> 
> Ha, you are now @paescuj! 😄

Hi Rijkvanzanten, sorry I did not see this sooner. 

It's possible I have not explored the full range of options (for example, putting CloudFront in front of Directus! Which I did not consider) but in my situation, after Directus uploads files (mostly image assets) to the cloud (in my case DigitalOcean Spaces), I have to kind of finesse the URL using the ID and the other environment variables to construct it if I want the CDN link. So I'm able to get the CDN link but it's not ideal, because the name of the file is just a UUID because Directus renames the image files. 

I don't want to serve the files over my server if I can help it and would prefer to use a CDN. Perhaps for the thumbnails it's not a big deal but I'd still rather just serve the links and save the bandwidth. 

Probably not relevant but I'm also not exposing the Directus admin panel publicly in my case (access is through a VPN tunnel only); I have a node.js server that interfaces with the Directus API so that part is not a big problem.  

Ideally, in the Directus admin panel I would have (optional) access to a CDN link for these images and possibly even generated thumbs right through calls to the Directus API, and the files would be nicely named for SEO etc. I understand why the files have to have UUIDs for names but in my case, there are not so many files that our naming convention would ever cause an issue, and I'd rather they just keep the name they started with. 

Perhaps I have not explored the full range of options, so let me know if you have suggestions! But in my case, permissions and naming, which are some of the draws of using Directus to get the assets, are actually not a concern (my assets will all be public for now), whereas SEO and the speed of a CDN are much more important. 

Thanks again for your response and let me know what you think.

**Reply by rijkvanzanten** [reply]
So just to be absolutely sure, in your use case it's ok that every file is publicly available / scrapeable? That's been one of the main parts here that's been tricky to resolve

**Reply by ardenschager** [reply]
> So just to be absolutely sure, in your use case it's ok that every file is publicly available / scrapeable? That's been one of the main parts here that's been tricky to resolve

In my case, yeah that is not an issue. It's actually nice because then they would end up on Google Images. I can understand situations where that would not be appealing but in the case of this current project the tradeoff is fine; if there was an option to choose somewhere in Directus between having stricter access permissions for these images vs just assuming the assets are exposed on a CDN, I would just expose all of the assets for the sake of bandwidth, SEO, convenience, and lack of a need to keep them otherwise hidden in this case.

**Reply by ardenschager** [reply]
I should also mention that one of the workarounds I have at the moment is to use imagekit attached to DigitalOcean Spaces to optimize/resize + serve thumbnails over a CDN. But it would be more convenient to handle this in Directus if that's ever possible.

**Reply by rijkvanzanten** [reply]
> I should also mention that one of the workarounds I have at the moment is to use imagekit attached to DigitalOcean Spaces to optimize/resize + serve thumbnails over a CDN

Directus already has an on-the-fly thumbnail generation endpoint :) https://docs.directus.io/reference/files.html#requesting-a-thumbnail Putting a CDN in front of Directus. Also if file-name SEO on your website is a concern, you can put any arbitrary text after the UUID in the assets endpoints for that exact SEO reason (like `/assets/<uuid>/my-custom-name.jpg` 👍🏻

**Reply by ardenschager** [reply]
> > I should also mention that one of the workarounds I have at the moment is to use imagekit attached to DigitalOcean Spaces to optimize/resize + serve thumbnails over a CDN
> 
> Directus already has an on-the-fly thumbnail generation endpoint :) https://docs.directus.io/reference/files.html#requesting-a-thumbnail Putting a CDN in front of Directus. Also if file-name SEO on your website is a concern, you can put any arbitrary text after the UUID in the assets endpoints for that exact SEO reason (like `/assets/<uuid>/my-custom-name.jpg` 👍🏻

Thanks for these tips. I was aware Directus can do thumbs but I was more concerned with serving them over a CDN. I would be open to experimenting with putting a CDN in front of Directus, if you have a sec can you let me know what that process would look like?

**Reply by rijkvanzanten** [reply]
Should be the same as any other CDN you'd use! You point the CDN's origin to your public Directus address, and you should be good to go 👍🏻

**Reply by ardenschager** [reply]
Awesome, thank you very much for your patience! I'm gonna play around with it more then and see if I can be less hacky and try to use Directus as intended :)

**Reply by M-GRV** [reply]
I could see this being useful for orgs that need to pay for API metering now and need some of their files served publicly like ecommerce sites.

**Reply by MarvinXu** [reply]
Hi @ardenschager , how's the solution work so far? It seems to me that it doesn't solve the bandwidth issue you mentioned above?

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by omid** [reply]
> The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

Thanks. Is it approved or denied?

**Reply by SalahAdDin** [reply]
> > The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.
> 
> Thanks. Is it approved or denied?

Same question here

**Reply by rijkvanzanten** [reply]
It's an open feature request and something we're looking to include in a future release 👍🏻

**Comment by mscbpi** [comment]
@rijkvanzanten 

How would you consider this two years later:
1. delivering content directly from cloud object storage, through a CDN for caching and scalability (considering an external API builds the asset URL based on Directus API answer)

2. delivering from directus (itself pulling from object storage) and putting a CDN in front of directus, caching image assets (and other objects if needed).

=>

1. would save directus ressources for delivering assets

2. would enable directus features when requesting the asset like image transformation (could be handled by some CDNs as well), more resource usage while still benefiting from a CDN.

I don't know what would be your best case scenario/ best practice architecture, the way you imagine directus being used for delivering assets.

**Reply by rijkvanzanten** [reply]
Number 2 is how it works now 👍🏻  Number 1 is tricky to implement reliably. F.e. an S3 bucket may or may not be publicly accessible, and may or may not have a CDN in front of it, whereas other supported storage locations like Cloudinary have a known structure.

**Reply by mscbpi** [reply]
edit: issue open there:
https://github.com/directus/directus/issues/20055

TL;DR:

CDN Caching (Azure Front Door) in front of Directus when using `azure` storage fails due to `range` requests failing on Directus in that case.

Same setup using `local` storage works well.

Same request to Blob storage works well (so does CDN for scenario 1.).

**Comment by readtedium** [comment]
As someone who is just coming to Directus as a potential opportunity and still a fairly new user, I just want to throw my hat in the ring to note some reasons direct linking to existing asset storage is a key feature:

- I have an existing archive of images that I serve through another service (S3 with Imgix in front), and to move them to Directus could potentially cost me years of built-up SEO.
- I’m moving asset content from another database into this one. Having to move URLs complicates this transition and makes it so I have to update a lot of assets manually, slowing down what could otherwise be a somewhat painless content migration.
- Sure, linking images manually from their existing sources would be possible, and I could potentially move to a new approach going forward. But I worry this just clutters the database with fields that shouldn’t be necessary.

To put this another way: As a test, I tried moving over my existing content using Directus’ built-in import tools, and found that process to be extremely painless for the content itself. Moving over thousands of posts, while not perfect of course, was relatively easy. But it feels like if you want to do the same thing with existing assets that you didn’t even intend to move from their current location, you run into roadblocks.

I understand there are edge cases that can emerge, as @rijkvanzanten noted in the comment thread above, but I also think that these are edge cases that those interested in this feature are knowingly signing up for.

I realize that others are also probably managing considerations like this, but I hope my perspective as someone who wants to transition a site with existing content to Directus is useful in considering next steps.

**Comment by geekyayush** [comment]
Any update on this?

**Comment by BorisKamp** [comment]
Following this and plus one-ing this as it's very important to me too

**Comment by xndyz** [comment]
This would be a super useful feature, as the cost for bandwidth can differ drastically serving it from a CDN on public storage URL vs Directus.

**Comment by pateljoel** [comment]
This feature is a must, I am currently transforming a Directus URL from /assets/<UUID> to my own CDN and it would be great if directus can use my CDN to insert an image into markdown

**Comment by clhey** [comment]
Four years have passed, is this feature available yet?

**Comment by eallion** [comment]
Need this!

Each API request takes a significant amount of time to convert file image URLs to the CDN domain and include the `filename_disk`.

**Comment by hofstettera** [comment]
This feature would also be very relevant for us.
I would be grateful if we could get an estimate of when or if we can expect it at some point.

**Comment by timstapl** [comment]
I've been using Directus for a couple of projects again, and have once again run into this issue. I'm going to take a look and see if I can get a pull request for the current version together, taking the feedback on my prior pull request into account.

**Comment by perakasem** [comment]
I'm Running into this issue using R2. Please give this feature request a nudge on the new roadmap: 

https://roadmap.directus.io/roadmap?id=5785688a-6a36-4c57-98fb-6816afeae04c
